### PR TITLE
Refactor shouldDelegateList/shouldListFromStorage to better explain d…

### PR DESCRIPTION
**Pull request by serathius**
/kind cleanup

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

Extracted from https://github.com/kubernetes/kubernetes/pull/118508

Helps https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/2340-Consistent-reads-from-cache

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go [hunk 1]
```diff
@@ -721,17 +721,17 @@ func shouldDelegateList(opts storage.ListOptions) bool {
...
 	match := opts.ResourceVersionMatch
 	pagingEnabled := utilfeature.DefaultFeatureGate.Enabled(features.APIListChunking)
+
+	// Serve consistent reads from storage
+	consistentReadFromStorage := resourceVersion == ""
+	// Watch cache doesn't support continuations, so serve them from etcd.
 	hasContinuation := pagingEnabled && len(pred.Continue) > 0
+	// Serve paginated requests about revision "0" from watch cache to avoid overwhelming etcd.
 	hasLimit := pagingEnabled && pred.Limit > 0 && resourceVersion != "0"
+	// Watch cache only supports ResourceVersionMatchExact (default).
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:731]: It's vice versa - it doesn't support ResourceVersionMatchExact.

Comment (serathius) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go]: Ups, would be great to remove negations to make it easier to comprehend. With the new tests it should be safe.

**Pull request merged**